spikesorters/ironclust/ironclust.py

```diff
@@ -1,159 +1,163 @@
-        # "p_ironclust('$(tempdir)','$timeseries$','$geom$','$prm$','$firings_true$','$firings_out$','$(argfile)');"
-        cmd_call = "p_ironclust('{}', '{}', '{}', '', '', '{}', '{}');" \
-            .format(output_folder, dataset_dir / 'raw.mda', dataset_dir / 'geom.csv', output_folder / 'firings.mda',
-                    dataset_dir / 'argfile.txt')
-        cmd = 'matlab -nosplash -nodisplay -r "{} {} quit;"'.format(cmd_path, cmd_call)
-        if self.debug:
-            print(cmd)
-
-        if "win" in sys.platform:
-            cmd_list = ['matlab', '-nosplash', '-nodisplay', '-wait',
-                        '-r', '{} {} quit;'.format(cmd_path, cmd_call)]
-        else:
-            cmd_list = ['matlab', '-nosplash', '-nodisplay',
-                        '-r', '{} {} quit;'.format(cmd_path, cmd_call)]
-
-        _call_command_split(cmd_list)
+        with open(dataset_dir / 'argfile.txt', 'w') as f:
+            f.write(txt)
+
+        tmpdir = output_folder / 'tmp'
+        os.mkdir(tmpdir)
+        print('Running ironclust in {tmpdir}...'.format(tmpdir=str(tmpdir)))
+        cmd = '''
+            addpath('{source_dir}');
+            addpath('{ironclust_path}', '{ironclust_path}/matlab', '{ironclust_path}/matlab/mdaio');
+            try
+                p_ironclust('{tmpdir}', '{dataset_dir}/raw.mda', '{dataset_dir}/geom.csv', '', '', '{tmpdir}/firings.mda', '{dataset_dir}/argfile.txt');
+            catch
+                fprintf('----------------------------------------');
+                fprintf(lasterr());
+                quit(1);
+            end
+            quit(0);
+        '''
+        cmd = cmd.format(ironclust_path=IronClustSorter.ironclust_path, tmpdir=str(tmpdir), dataset_dir=str(dataset_dir), source_dir=str(source_dir))
+
+        matlab_cmd = ShellScript(cmd, script_path=str(tmpdir / 'run_ironclust.m'), keep_temp_files=True)
+        matlab_cmd.write()
+
+        shell_cmd = '''
+            #!/bin/bash
+            cd {tmpdir}
+            matlab -nosplash -nodisplay -r run_ironclust
+        '''.format(tmpdir=str(tmpdir))
+        shell_script = ShellScript(shell_cmd, script_path=str(tmpdir / 'run_ironclust.sh'), keep_temp_files=True)
+        shell_script.start()
+
+        retcode = shell_script.wait()
+
+        if retcode != 0:
+            raise Exception('ironclust returned a non-zero exit code')
+
+        result_fname = str(tmpdir / 'firings.mda')
+        if not os.path.exists(result_fname):
+            raise Exception('Result file does not exist: ' + result_fname)
+
+        samplerate_fname = str(tmpdir / 'samplerate.txt')
+        with open(samplerate_fname, 'w') as f:
+            f.write('{}'.format(samplerate))
 
     @staticmethod
-    def get_result_from_folder(output_folder):
-
-        # overwrite the SorterBase.get_result
-        from mountainlab_pytools import mdaio
+    def get_result_from_folder(output_folder: Path):
```

Comment:
with (output_folder: Path) the naive user not aware of this Path object will be force to use it. No ?
Could we also accept str and convert it to Path ?

Reply:
This is why I think it's important to static type the entire project. Did you see that issue I raised on SpikeExtractors? If it can be Path or str, then we need to know. I use pyright which is a superb tool for vscode and I am catching a lot of errors before they even have a chance to cause exceptions.

Reply:
shoulde we use Union[str, Path]?

Reply:
It should be the same type throughout the entire project... so the question is... do you want to handle both cases throughout? Personally I would stick to one.

Reply:
I think it's fine to handle both. Just parse to Path whatever input you get and it works

Reply:
That's okay. But thus we see the importance of static typing (with Union[str, Path] in this case).... because the linter will warn you if you are not making the proper conversion. Otherwise there will be bugs that only happen in particular settings.

Reply:
@alejoe91 ... just do a search for the following text: "output_folder / "

All those cases will presently fail if a str is passed in.